gtsfm/ui/process_graph_generator.py

```diff
@@ -0,0 +1,167 @@
+        # http://robertyu.com/wikiperdido/Pydot%20Clusters
+        self._plate_to_cluster = {}
+
+        self._test_mode = test_mode
+
+    def _build_graph(self) -> None:
+        """Build graph based on RegistryHolder's REGISTRY."""
+
+        unique_metadata = self._get_metadata_from_registry()
+
+        # sort list to prevent non-deterministic DOT graph
+        for metadata in sorted(list(unique_metadata)):
+            self._add_nodes_to_graph(metadata)
+
+        for plate_name, cluster in self._plate_to_cluster.items():
+            self._main_graph.add_subgraph(cluster)
+
+    def _get_metadata_from_registry(self) -> Set[UiMetadata]:
+        """Get a set of unique_metadata from the central registry.
+
+        Also, create empty pydot clusters for later use as plate subgraphs, and store in self._plate_to_cluster.
+        """
+        unique_metadata = set()
+
+        for cls_name, cls_type in RegistryHolder.get_registry().items():
+            # don't add the base class to the graph
+            if cls_name == "GTSFMProcess":
+                continue
+
+            # don't add any test classes to the graph, unless in testing mode
+            if not self._test_mode and cls_name.startswith("Fake"):
+                continue
+
+            # get UI metadata of class
+            metadata = cls_type.get_ui_metadata()
+
+            # skip duplicates
+            # happens when concrete classes implement an abstract class without overwriting get_ui_metadata()
+            if metadata in unique_metadata:
+                continue
+            unique_metadata.add(metadata)
+
+            # create an empty plate for each unique parent_plate name
+            plate = metadata.parent_plate
+            if plate not in self._plate_to_cluster:
+                # if no plate, add straight to main graph
+                if plate == "":
+                    continue
+
+                new_cluster = pydot.Cluster(plate, label=plate)
+                self._plate_to_cluster[plate] = new_cluster
+
+        return unique_metadata
+
+    def _add_nodes_to_graph(self, metadata: UiMetadata) -> None:
+        """Add connected blue/gray nodes to relevant subgraph based on UiMetadata.
+
+        Also, auto-casts metadata.input_products and metadata.output_products to tuples of strings.
+
+        Why? A common error is to define a one-element tuple like so:
+          > input_products = ("Input Product")
+        but this is actually parsed as a raw string. The correct usage is:
+          > input_products = ("Input Product",)
+        which is unintuitive. Auto-cast here prevents unexpected behavior for developers.
+
+        Args:
+            metadata: UiMetadata object to add nodes/edges for
+        """
+
+        display_name = metadata.display_name
+        style = self._style
+
+        # autocast strings to one-element tuples
+        input_products = metadata.input_products
+        if type(input_products) == str:
```

Comment:
Would it be cleaner to use a fixed type for `input_products`, e.g. always just return a Tuple?

Reply:
Yes. But I ran into this issue with one-element tuples (described in lines 115-119 [above](https://github.com/borglab/gtsfm/pull/540/files/c7ae439db14f58a05e7754a8707bac0ee9dd136d#diff-ce2386876f8582f4228575c840a96fc7350133a79df6a912b9b2bd2017020fdbR115-R119)). 

Since Python is not statically typed it can be easy to accidentally return a string instead of a one element tuple, which is why I added this autocast here. I can remove it but I feel that would be sacrificing a useful guardrail in pursuit of clean code.